session/schema_amender.go

```diff
@@ -128,6 +130,72 @@ func (a *amendCollector) keyHasAmendOp(key []byte) bool {
 	return len(ops) > 0
 }
 
+func needCollectIndexOps(actionType uint64) bool {
+	return actionType&(1<<model.ActionAddIndex) != 0
+}
+
+func needCollectModifyColOps(actionType uint64) bool {
+	return actionType&(1<<model.ActionModifyColumn) != 0
+}
+
+func fieldTypeDeepEquals(ft1 *types.FieldType, ft2 *types.FieldType) bool {
+	if ft1.Tp == ft2.Tp &&
+		ft1.Flag == ft2.Flag &&
+		ft1.Flen == ft2.Flen &&
+		ft1.Decimal == ft2.Decimal &&
+		ft1.Charset == ft2.Charset &&
+		ft1.Collate == ft2.Collate &&
+		len(ft1.Elems) == len(ft2.Elems) {
+		for i, elem := range ft1.Elems {
+			if elem != ft2.Elems[i] {
+				return false
+			}
+		}
+		return true
+	}
+	return false
+}
+
+// colChangeAmendable checks whether the column change is amendable, now only increasing column field
+// length is allowed for committing concurrent pessimistic transactions.
+func colChangeAmendable(colAtStart *model.ColumnInfo, colAtCommit *model.ColumnInfo) error {
+	// Modifying a stored generated column is not allowed by DDL, the generated related fields are not considered.
+	if !fieldTypeDeepEquals(&colAtStart.FieldType, &colAtCommit.FieldType) {
+		if colAtStart.FieldType.Flag != colAtCommit.FieldType.Flag {
+			return errors.Trace(errors.Errorf("flag is not matched for column=%v, from=%v to=%v",
+				colAtCommit.Name.String(), colAtStart.FieldType.Flag, colAtCommit.FieldType.Flag))
+		}
+		if colAtStart.Charset != colAtCommit.Charset || colAtStart.Collate != colAtCommit.Collate {
+			return errors.Trace(errors.Errorf("charset or collate is not matched for column=%v", colAtCommit.Name.String()))
+		}
+		err := ddl.CheckModifyTypeCompatible(&colAtStart.FieldType, &colAtCommit.FieldType)
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+	// TODO Default value change is not supported.
+	if !reflect.DeepEqual(colAtStart.DefaultValue, colAtCommit.DefaultValue) {
```

Comment:
Is it possible that the default value changes from oldDefaultValue to newDefaultValue and then to oldDefaultValue？

Reply:
It's ok if they are the same at last, but I'm not sure if it'ok to check default value change in this way.

Reply:
Yes, even if `modify column c (new defaultValue)` -> `add index(c)` -> `modify column c (old defaultValue)`. Becuase these operations all in pessimistic transactions.

Reply:
Could we add a test like `modify column c (new defaultValue)` -> `add index(c)` -> `modify column c (old defaultValue)`？
Add a test like the default value is changed to `now()`?